c-api/compat-5.3.c

```diff
@@ -336,23 +372,134 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
 
 
 COMPAT53_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
+  const char* s = NULL;
   int en = errno;  /* calls to Lua API may change this value */
+#if (defined(COMPAT53_HAVE_STRERROR_R) && COMPAT53_HAVE_STRERROR_R) || (defined(COMPAT53_HAVE_STRERROR_S) && COMPAT53_HAVE_STRERROR_S)
+  char buf[512] = { 0 };
+#endif /* buffer for threadsafe variants of strerror if possible */
   if (stat) {
     lua_pushboolean(L, 1);
     return 1;
   }
   else {
     lua_pushnil(L);
+#if (defined(COMPAT53_HAVE_STRERROR_R) && COMPAT53_HAVE_STRERROR_R)
+    /* use strerror_r here, because it's available on these specific platforms */
+#if defined(COMPAT53_HAVE_STRERROR_R_XSI)
+    /* XSI Compliant */
+    strerror_r(en, buf, 512);
+    s = buf;
+#else
+    /* GNU-specific which returns const char* */
+    s = strerror_r(en, buf, 512);
+#endif
+#elif (defined(COMPAT53_HAVE_STRERROR_S) && COMPAT53_HAVE_STRERROR_S)
+    /* for MSVC and other C11 implementations, use strerror_s 
+     * since it's provided by default by the libraries 
+     */
+    strerror_s(buf, 512, en);
+    s = buf;
+#else
+    /* fallback, but
+     * strerror is not guaranteed to be threadsafe due to modifying
+     * errno itself and some impls not locking a static buffer for it
+     * ... but most known systems have threadsafe errno: this might only change
+     * if the locale is changed out from under someone while this function is being called
+     */ 
+    s = strerror(en);
+#endif
     if (fname)
-      lua_pushfstring(L, "%s: %s", fname, strerror(en));
+      lua_pushfstring(L, "%s: %s", fname, s);
     else
-      lua_pushstring(L, strerror(en));
+      lua_pushstring(L, s);
     lua_pushnumber(L, (lua_Number)en);
     return 3;
   }
 }
 
 
+static const char* compat53_f_parser_handler (lua_State *L, void *data, size_t *size) {
+  (void)L; /* better fix for unused parameter warnings */
+  struct compat53_f_parser_buffer *p = (struct compat53_f_parser_buffer*)data;
+  size_t readcount = fread(p->buffer + p->start, sizeof(*p->buffer), sizeof(p->buffer), p->file);
+  if (ferror(p->file) != 0 || feof(p->file) != 0) {
```

Comment:
We should only check for error and/or eof, if we can't read a single byte. Otherwise we return what we got.

Reply:
I restructured the code to read 1 byte (if possible), check for errors, proceed if there are none. We then check for `feof` at the beginning now of this handler function. If reading 1 byte returned EOF, we exit immediately (`feof` tells us to). Otherwise, we read and then check for errors (which does not include EOF): if were successful, we pass that back to Lua to handle. It then continues this pattern and appropriately consumes the whole file (I went through and checked a reading of a few files, just to make sure, and it works appropriately and sol2's tests pass).